langchain/text_splitter.py

```diff
@@ -64,10 +64,15 @@ def create_documents(
                 documents.append(new_doc)
         return documents
 
-    def split_documents(self, documents: List[Document]) -> List[Document]:
+    # can be an iterable or a sequence
+    def split_documents(
+        self, documents: Union[Iterable[Document], Sequence[Document]]
+    ) -> List[Document]:
         """Split documents."""
-        texts = [doc.page_content for doc in documents]
-        metadatas = [doc.metadata for doc in documents]
+        texts, metadatas = [], []
```

Comment:
Could you provide a benchmark for this?

Since list comprehensions are fast and this code involves mutating lists, I don't know if it's obvious what version will be fater:

```python
a = [ (x, x+1) for x in range(int(1e7)) ]

x = []
y = []

for z in a:
    x.append(z[0])
    y.append(z[1])
```

```python
a = [ (x, x+1) for x in range(int(1e7)) ]

x = [ z[0] for z in a ]
y = [ z[1] for z in a ]
```

Reply:
@eyurtsev See long form response below. There is a much faster way to do this, but the way I proposed in this PR is no slower.

Reply:
@eyurtsev See below for a benchmark analysis across different document lengths.

Reply:
In summary, it's no slower the way it's proposed. The primary advantage of the second function (proposed in this PR) is _not_ the incremental improvement for smaller length sequences of Document, but rather that it doesn't need to iterate over the documents twice, allowing for effective usage of Python generators.

Reply:
@eyurtsev 

Below is an image of the speedup for "meatier" `page_content` and `metadata` attributes to provide more confidence on this speedup. The associated notebook generates that image. Langchain is not required to run this.

![image](https://github.com/hwchase17/langchain/assets/689723/a8ec6bdd-48d1-44de-b71f-691d9e8bc943)
[benchmark_for_text_metadata_split.pdf](https://github.com/hwchase17/langchain/files/11536857/benchmark_for_text_metadata_split.pdf)

Reply:
> but rather that it doesn't need to iterate over the documents twice,

Apologies I should've read the PR summary carefully!